src/ast.ts

```diff
@@ -76,10 +70,9 @@ export class AST {
     // When an AST is to be used by CMB, it must be annotated.
     // This step is computationally intensive, and in certain instances
     // unecessary
-    if (annotate) this.annotateNodes();
-
-    this.id = -1; // just for the sake of having an id, though unused
-    this.hash = hashObject(this.rootNodes.map((node) => node.hash));
+    if (annotate) {
+      this.annotateNodes();
+    }
```

Comment:
I left this here, but I can't find any place in the code where we set `annotate=false`. Can I remove the conditional check?

Reply:
The pretty-printer normalizes the sourcecode according to various rules, such that future parse-print-parse cycles are all nondestructive. But during the _first_ parse, this can result in changes that affect location information.

So when switching from Text to Blocks, we need to parse _twice_. First, [we parse to a CMB AST](https://github.com/bootstrapworld/codemirror-blocks/blob/master/src/ui/ToggleEditor.tsx#L541). If the parse is successful, we use the .toString() method to [pretty-print back to text](https://github.com/bootstrapworld/codemirror-blocks/blob/master/src/ui/ToggleEditor.tsx#L553), and then [parse _again_.](https://github.com/bootstrapworld/codemirror-blocks/blob/master/src/ui/ToggleEditor.tsx#L554)

Annotating a large program is pretty expensive, so my goal is to only annotate the second parse. I haven't switched this on yet, but that's where I'd be passing `false`.

Reply:
I see. My inclination is to remove it since it's not currently being used, but it would be good to have a plan for benchmarking that's documented somewhere. In an ideal world, code that is only there for performance would be tested via some automated benchmarking.

Reply:
It's dead code, and dead code should always be removed unless it's brand new code that just hasn't been hooked up yet and is about to be hooked up imminently, in which case there should probably be tests for it.

In this particular case, it's hard to tell looking at the code that it's not actually meant to do anything, but is actually just a comment that says 

```
// TODO: implement performance optimizations and benchmarking
```

Of course people also have strong opinions about the appropriateness of leaving `// TODO` comments in the code instead of in an issue tracker, but that's more of a project management question.

In any case, it's kind of a "death by a thousand cuts" problem. A little bit of dead code here and there isn't so bad, but the more you allow dead code to persist, the more dead code you end up with, until eventually you have so much dead code that any kind of refactoring or feature additions become very time consuming. It's like the garbage collector problem. If you free up the memory you've used as soon as you're done with it, it costs a bit more compute up front. But then you don't have to worry about long GC pauses later that bring your program to a screeching halt while all your memory has to be shuffled around. If you remove the dead code as soon as it's no longer used, it takes a bit more work up front but you don't have to worry about doing a big complicated refactor later to figure out and remove all the dead code before continuing to develop new features.